drt_1d.py

```diff
@@ -8,121 +8,289 @@
+        self.widths = np.array([region.width for region in regions])
+        self.sigmas = np.array([region.scattering_cross_section for region in regions])
+        self.num_voxels = sum(region.num_voxels for region in self.regions)
+        self.cumsum_voxels = np.cumsum([0] + [region.num_voxels for region in self.regions])
+        self._compute_boundaries()
+        
+    def _compute_boundaries(self) -> np.ndarray:
+        """
+        Computes the boundary positions of each region within the domain. 
+        
+        Includes starting and ending edges of the domain.
+
+        Returns:
+        - np.ndarray: Array of boundary positions.
+        """
+        self.boundaries = np.cumsum([self.start_x] + [region.width for region in self.regions])
+        for start_x, region in zip(self.boundaries[:-1],self.regions):
+            region.start_x = start_x
+
+    def get_region_id_from_position(self, x:float) -> int:
+        """
+        Determines the region ID for a given x-coordinate.
+        
+        Note: np.searchsorted uses binary search. 
+        Additionally, this computation includes the left edge in the search but not the right edge.
+
+        Parameters:
+        - x (float): x-coordinate.
+
+        Returns:
+        - int: Region ID containing the x-coordinate.
+        """
+        return np.searchsorted(self.boundaries, x, side='right') - 1
     
-    s = np.zeros(num_voxels)
-    s[source_voxel_index] = source_strength
-    logging.debug(f"Constructed source vector: {s}")
-    return s
+    def get_region_from_voxel_index(self, voxel_index: int) -> int:
+        """
+        Determines the region ID for a given voxel index.
 
-def calculate_K_trans(ri:int, rj:int, voxel_size:float, sigma_s:np.ndarray):
-    """
-    Calculate the streaming operator. 
+        Parameters:
+        - voxel_index (int): Index of the voxel.
+
+        Returns:
+        - int: Region ID containing the voxel.
+        """
+        return np.searchsorted(self.cumsum_voxels, voxel_index, side='right') - 1
     
-    Used by the construct_transition_matrix function to build the K_trans matrix.
+    def get_voxel_id(self, x: float) -> int:
+        """
+        Determines the voxel ID for a given x-coordinate.
+
+        Note: an error is thrown when the given x-coordinate is out of bounds or on the right boundary.
+
+        Parameters:
+        - x (float): x-coordinate.
 
-    Current method calculates tau by summing the cross-sections of the voxels from the start index to the end index, 
-    always summing from the left edge of the starting voxel to the left edge of the ending voxel, regardless of the direction of transport.
+        Returns:
+        - int: Voxel ID containing the x-coordinate.
+        """
+        region_idx = self.get_region_id_from_position(x)
+        region = self.regions[region_idx]
+        local_x = x - self.boundaries[region_idx]
+        voxel_id_within_region = int(np.floor(local_x / region.voxel_width))
```

Comment:
I think there is value in making this a method of the `Region` class:
```
    def get_local_voxel_id(self, x):
        return int(np.floor( (x - self.start_x) / region.voxel_width))
```

Bonus: if you store the first voxel id for each region in the same way you store the start x for each region, at initialization, you could include that here too  

```
    def get_global_voxel_id(self, x):
        return self.get_local_voxel_id(x) + self.first_voxel_id
```